### PR TITLE
Log collecting and query implementation

### PR DESCRIPTION
**Pull request by arugal**
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

https://github.com/apache/skywalking/issues/5930

### CODE REVIEW

**Review by wu-sheng — COMMENTED** [review]
Note, your skywalking-ui submodule is reverting the commit.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/TraceAnalyzer.java
Comment (wu-sheng) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/TraceAnalyzer.java:75]: Why do you remove try/catch? Just expose the error to handler?

Comment (arugal) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/TraceAnalyzer.java:75]: https://github.com/apache/skywalking/blob/f54f639c4d9207c20490519121eeabe6acf37374/oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/handler/v8/grpc/TraceSegmentReportServiceHandler.java#L72-L78

> Just expose the error to handler?

Yes, otherwise `errorCounter` is incorrect.

Comment (wu-sheng) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/TraceAnalyzer.java:75]: Remember to add this kind of fix(bug) in the changlog, which you have submitted yet.

File: oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/listener/TrafficAnalysisListener.java
Comment (wu-sheng) [inline oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/listener/TrafficAnalysisListener.java:72]: You seems assuming there will be service name in the log? I would say maybe not. We should add a check to ignore this kind of logs in the beginning.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/AbstractLog.java
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/AbstractLog.java:32]: Should be `timebucket`, I think.

Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/AbstractLog.java:32]: `abstract class Source` includes this already.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/EndpointMeta.java
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/EndpointMeta.java:28]: Why need this? We have `EndpointTraffic` for the metadata of endpoint. What do I miss?

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/EndpointMeta.java:28]: OK, I think I get it now.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/EndpointMeta.java
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/EndpointMeta.java:30]: Typically, I think we usually have 
```
    @ScopeDefaultColumn.DefinedByField(columnName = "name", requireDynamicActive = true)
    private String name;

    @Setter
    private NodeType serviceNodeType;
```

I know this would cause duplicated encoding, but this could support endpoint of any node type in the future. We did this in the `EndpointRelation` too.

File: oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/LogQueryEsDAO.java
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/LogQueryEsDAO.java:99]: Don't forget this. Seems existing in all `ILogQueryDAO` implementations.

File: oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/h2/dao/H2LogQueryDAO.java
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/h2/dao/H2LogQueryDAO.java:64]: This constructor is never being used. Please recheck. Currently, the provider is using `public H2LogQueryDAO(JDBCHikariCPClient h2Client) `.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/AbstractLog.java
Comment (arugal) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/AbstractLog.java]: done

File: oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/h2/dao/H2LogQueryDAO.java
Comment (arugal) [reply oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/h2/dao/H2LogQueryDAO.java:64]: Thank. I'll fix it

File: oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/LogAnalyzer.java
Comment (wu-sheng) [inline oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/LogAnalyzer.java:49]:
```suggestion
            // If no timestamp, OAP server would use the received timestamp as log's timestamp
```

File: oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/listener/TrafficAnalysisListener.java
Comment (arugal) [reply oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/listener/TrafficAnalysisListener.java:67]: done

File: oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/LogQueryEsDAO.java
Comment (arugal) [reply oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/LogQueryEsDAO.java]: done

**Comment by EvanLjp** [comment]
LGTM

**Comment by wu-sheng** [comment]
Link https://github.com/apache/skywalking-query-protocol/pull/46, as this PR would include the implementation.

File: oap-server/server-receiver-plugin/skywalking-log-recevier-plugin/src/main/java/org/apache/skywalking/oap/server/recevier/log/provider/handler/LogReportServiceHandler.java
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-log-recevier-plugin/src/main/java/org/apache/skywalking/oap/server/recevier/log/provider/handler/LogReportServiceHandler.java:71]: Please note this at the protocol level, https://github.com/apache/skywalking-data-collect-protocol/blob/345d9edac842179e61504999c91bf2bec6ae0ff7/logging/Logging.proto#L50

Also, this only works as the streaming mode in the gRPC, or batch mode of HTTP receivers. It isn't suitable for the queue.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:85]: If this `storageOnly = true`, we should consider `binary` rather than `string`. 
At the same time, depend on your suggestions here, https://github.com/apache/skywalking-query-protocol/pull/47

**Comment by wu-sheng** [comment]
@arugal The query protocol updated. Please follow that.

https://github.com/apache/skywalking-query-protocol/blob/4ac87d563213a2ff70039afc0b7a5a4815878bbd/log.graphqls#L58-L61

File: .github/workflows/e2e.log.yaml
Comment (arugal) [reply .github/workflows/e2e.log.yaml]: Thx, fixed

File: oap-server/server-receiver-plugin/skywalking-log-recevier-plugin/src/main/java/org/apache/skywalking/oap/server/recevier/log/provider/handler/LogReportServiceHandler.java
Comment (arugal) [reply oap-server/server-receiver-plugin/skywalking-log-recevier-plugin/src/main/java/org/apache/skywalking/oap/server/recevier/log/provider/handler/LogReportServiceHandler.java]: done

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:103]: I am a little confused about `dataBinary`, `tags`, and `tagsRawData`.
I could understand 2 of them should be raw data and search tags, then what is the 3rd?

Comment (arugal) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:103]: The `dataBinary` is the raw data for all the tags, `tags` and `tagsRawData` are search tags. `tagsRawData` is used by H2 and MySQL storage, I should rename the `tags` and `tagsRawData` to `searchTags` and `searchTagsRawData`

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:111]: Take a look at a thing called `DataTable`? It is supported in all storages, and you also could implement your own `StorageDataComplexObject` to support multiple storage options.
Don't provide duplicated field which causes confusing.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:115]: Typically, we don't use the un-persistent-able type as the field type of this kind of core concept.

Comment (arugal) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:115]: `tagsRawData` is converted to a single column store in H2 and MySQL, `StorageDatabacomplexObjec` does not apply here ?

https://github.com/apache/skywalking/blob/688a9c7b057ac82b6fef233a1df184d9c21867cc/oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/h2/dao/H2SegmentRecordBuilder.java#L95-L122

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:115]: Could you covert it to a `TagCollection` object, referencing the `DataTable` implementation? It would be better than you do this in every storage implementation.

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:115]: > tagsRawData is converted to a single column store in H2 and MySQL, StorageDatabacomplexObjec does not apply here ?

I am not saying you have to use that, <string, string> is not the DataTable for. But the suggestion is, making the persistent logic consistent in one place.
Like `KeyValueCollection` is also a good abstract name for this kind of value, right?

Comment (arugal) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java]: The fields have been renamed to `tagsRawData`, `tags` and `tagsInString`

**Review by wu-sheng — DISMISSED** [review]
Generally good to me. We could do more polish and enhancement later in the following PR.

@kezhenxu94 @dmsolr @EvanLjp Please recheck.

**Comment by wu-sheng** [comment]
E2E passed. Let's wait for other reviews' opinions.

**Review by kezhenxu94 — APPROVED** [review]
LGTM

**Review by kezhenxu94 — DISMISSED** [review]

**Comment by kezhenxu94** [comment]
Also, please add the 2 GitHub Actions files into https://github.com/apache/skywalking/blob/master/.github/workflows/cancel.yaml

**Comment by arugal** [comment]
> Also, please add the 2 GitHub Actions files into https://github.com/apache/skywalking/blob/master/.github/workflows/cancel.yaml

done

**Review by kezhenxu94 — DISMISSED** [review]

File: oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/LogQueryEsDAO.java
Comment (EvanLjp) [inline oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/LogQueryEsDAO.java:147]: desc is better than asc

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java
Comment (EvanLjp) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:90]: is there a way to add a custom analyzer?

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:90]: I will follow this later. Let's separate it as an enhancement.

File: oap-server/server-storage-plugin/storage-elasticsearch7-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch7/query/LogQueryEs7DAO.java
Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-elasticsearch7-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch7/query/LogQueryEs7DAO.java:149]: I think we should add order in LogQueryCondition at the protocol level.

Comment (arugal) [reply oap-server/server-storage-plugin/storage-elasticsearch7-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch7/query/LogQueryEs7DAO.java:149]: Add asc and desc to the query condition ?

Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-elasticsearch7-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch7/query/LogQueryEs7DAO.java:149]: Yes. a condition called order?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java
Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:90]: @EvanLjp Submit an issue to track this? I think we will have time to finish this in 8.4.0 milestone. Or do you want to do this?

Comment (EvanLjp) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/log/AbstractLogRecord.java:90]: no problem

File: oap-server/server-storage-plugin/storage-elasticsearch7-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch7/query/LogQueryEs7DAO.java
Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-elasticsearch7-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch7/query/LogQueryEs7DAO.java:149]: @arugal We have this already
```graphql
enum Order {
    ASC
    DES
}
```

Comment (arugal) [reply oap-server/server-storage-plugin/storage-elasticsearch7-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch7/query/LogQueryEs7DAO.java]: @EvanLjp @wu-sheng done, please recheck.

**Review by wu-sheng — APPROVED** [review]
LGTM. Once e2e are good, I am going to merge this version first. If there are something we need to enhance, we could do it later. There is still time before the 8.4.0 release due to agent plugin and UI side works.

**Pull request merged**
